native/dlaexc.go

```diff
@@ -0,0 +1,259 @@
+			return false
+		}
+
+		// Accept swap: apply transformation to the entire matrix T.
+		impl.Dlarfx(blas.Right, j3+1, 3, u[:], tau, t[j1:], ldt, work)
+		impl.Dlarfx(blas.Left, 3, n-j1-1, u[:], tau, t[j1*ldt+j2:], ldt, work)
+
+		t[j1*ldt+j1] = t33
+		t[j2*ldt+j1] = 0
+		t[j3*ldt+j1] = 0
+
+		if wantq {
+			// Accumulate transformation in the matrix Q.
+			impl.Dlarfx(blas.Right, n, 3, u[:], tau, q[j1:], ldq, work)
+		}
+
+	default: // n1 == 2 && n2 == 2
+		// Generate elementary reflectors H(1) and H(2) so that:
+		//  H(2) H(1) (  -X11  -X12 ) = (  *  * )
+		//            (  -X21  -X22 )   (  0  * )
+		//            ( scale    0  )   (  0  0 )
+		//            (    0  scale )   (  0  0 )
+		u1 := [3]float64{1, -x[ldx], scale}
+		_, tau1 := impl.Dlarfg(3, -x[0], u1[1:], 1)
+
+		temp := -tau1 * (x[1] + u1[1]*x[ldx+1])
+		u2 := [3]float64{1, -temp * u1[2], scale}
+		_, tau2 := impl.Dlarfg(3, -temp*u1[1]-x[ldx+1], u2[1:], 1)
+
+		// Perform swap provisionally on diagonal block in D.
+		impl.Dlarfx(blas.Left, 3, 4, u1[:], tau1, d[:], ldd, work)
+		impl.Dlarfx(blas.Right, 4, 3, u1[:], tau1, d[:], ldd, work)
+		impl.Dlarfx(blas.Left, 3, 4, u2[:], tau2, d[ldd:], ldd, work)
+		impl.Dlarfx(blas.Right, 4, 3, u2[:], tau2, d[1:], ldd, work)
+
+		// Test whether to reject swap.
+		m1 := math.Max(math.Abs(d[2*ldd]), math.Abs(d[2*ldd+1]))
+		m2 := math.Max(math.Abs(d[3*ldd]), math.Abs(d[3*ldd+1]))
+		if math.Max(m1, m2) > thresh {
+			return false
+		}
+
+		// Accept swap: apply transformation to the entire matrix T.
+		j4 := j1 + 3
+		impl.Dlarfx(blas.Left, 3, n-j1, u1[:], tau1, t[j1*ldt+j1:], ldt, work)
+		impl.Dlarfx(blas.Right, j4+1, 3, u1[:], tau1, t[j1:], ldt, work)
+		impl.Dlarfx(blas.Left, 3, n-j1, u2[:], tau2, t[j2*ldt+j1:], ldt, work)
+		impl.Dlarfx(blas.Right, j4+1, 3, u2[:], tau2, t[j2:], ldt, work)
+
+		t[j3*ldt+j1] = 0
+		t[j3*ldt+j2] = 0
+		t[j4*ldt+j1] = 0
+		t[j4*ldt+j2] = 0
+
+		if wantq {
+			// Accumulate transformation in the matrix Q.
+			impl.Dlarfx(blas.Right, n, 3, u1[:], tau1, q[j1:], ldq, work)
+			impl.Dlarfx(blas.Right, n, 3, u2[:], tau2, q[j2:], ldq, work)
+		}
+	}
+
+	if n2 == 2 {
+		// Standardize new 2×2 block T11.
+		a, b := t[j1*ldt+j1], t[j1*ldt+j2]
+		c, d := t[j2*ldt+j1], t[j2*ldt+j2]
+		a, b, c, d, _, _, _, _, cs, sn := impl.Dlanv2(a, b, c, d)
+		t[j1*ldt+j1], t[j1*ldt+j2] = a, b
+		t[j2*ldt+j1], t[j2*ldt+j2] = c, d
+		if j1+2 < n {
+			bi.Drot(n-j1-2, t[j1*ldt+j1+2:], 1, t[j2*ldt+j1+2:], 1, cs, sn)
+		}
+		if j1 > 0 {
+			bi.Drot(j1, t[j1:], ldt, t[j2:], ldt, cs, sn)
+		}
+		if wantq {
+			bi.Drot(n, q[j1:], ldq, q[j2:], ldq, cs, sn)
+		}
+	}
+	if n1 == 2 {
+		// Standardize new 2×2 block T22.
+		j3 := j1 + n2
+		j4 := j3 + 1
+		a, b := t[j3*ldt+j3], t[j3*ldt+j4]
+		c, d := t[j4*ldt+j3], t[j4*ldt+j4]
+		a, b, c, d, _, _, _, _, cs, sn := impl.Dlanv2(a, b, c, d)
+		t[j3*ldt+j3], t[j3*ldt+j4] = a, b
+		t[j4*ldt+j3], t[j4*ldt+j4] = c, d
+		if j3+2 < n {
```

Comment:
I think this is nicer as n - j3 - 2 > 0. Makes the mapping to the next line a lot clearer.